Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -266,27 +266,71 @@ public class GraphTraverser: GraphTraversing {
             test: isDependencyPrecompiledDynamicAndLinkable,
             skip: canDependencyEmbedProducts
         )
-        // Skip merged precompiled libraries from merging into the runnable binary
-        if case let .manual(dependenciesToMerge) = target.target.mergedBinaryType {
-            precompiledFrameworks = precompiledFrameworks
-                .filter { !isXCFrameworkMerged(dependency: $0, expectedMergedBinaries: dependenciesToMerge) }
-        }
-        references.formUnion(precompiledFrameworks.lazy.compactMap { self.dependencyReference(
-            to: $0,
-            from: .target(name: name, path: path)
-        ) })
-
         /// Other targets' frameworks.
         var otherTargetFrameworks = filterDependencies(
             from: .target(name: name, path: path),
             test: isDependencyDynamicTarget,
             skip: canDependencyEmbedProducts
         )
 
+        // Skip merged precompiled libraries from merging into the runnable binary first
+        switch target.target.mergedBinaryType {
+        case .automatic:
+            precompiledFrameworks = precompiledFrameworks
+                .filter { !isXCFrameworkMerged(dependency: $0) }
+        case let .manual(dependenciesToMerge):
+            precompiledFrameworks = precompiledFrameworks
+                .filter { !isXCFrameworkMerged(dependency: $0, expectedMergedBinaries: dependenciesToMerge) }
+        case .disabled:
+            break
+        }
+
+        // If runnable binary is a merged binary, keep only non-mergeable targets
         if target.target.mergedBinaryType != .disabled {
             otherTargetFrameworks = otherTargetFrameworks.filter(isDependencyDynamicNonMergeableTarget)
         }
 
+        // Search for merged non-runnable binaries
+        for innerTarget in otherTargetFrameworks {
+            guard let graphTarget = self.target(from: innerTarget),
+                  graphTarget.target.mergedBinaryType != .disabled
+            else {
+                continue
+            }
+
+            // Take all mergeable dependencies
+            let mergeableFrameworks = filterDependencies(
+                from: innerTarget,
+                test: isDependencyDynamicMergeableTarget,
+                skip: canDependencyEmbedProducts
+            )
+            switch graphTarget.target.mergedBinaryType {
+            case .automatic:
+                // Filter all dependencies from lists of precompiled binaries and local targets
+                precompiledFrameworks = precompiledFrameworks
```

Comment:
Shouldn't we filtering out from `precompiledFrameworks` direct and transitive pre-compiled dependencies of `innerTarget`? Or am I missing anything?

Reply:
Thanks, @pepicrft. That's an interesting question. From what I understand, if you have the following setup

```
- App 
    -> SharedFramework
- SharedFramework, dynamic
    -> frameworkA: dynamic, mergeable
    -> xcframeworkB: dynamic, mergeable
    -> xcframeworkC: dynamic, mergeable
```
you have 3 scenarios:

1. `SharedFramework` is not a merged binary
    - `frameworkA`, `xcframeworkB` and `xcframeworkC` should be embedded into the app
2. `SharedFramework` is a merged binary, marked as `automatic`
    - `frameworkA`, `xcframeworkB` and `xcframeworkC` should not be embedded into the app, because they are merged into the `SharedFramework`
4. `SharedFramework` is a merged binary, marked as `manual("xcframeworkC")`
    - `frameworkA`, `xcframeworkB` should be embedded into the app, while `xcframeworkC` - not, because it was defined that it should be merged.
    
Please correct me if you have any other knowledge on that, but that's what came up from all the docs and manipulations that I had with mergeable frameworks :)

Reply:
Overall, it is possible that the Project is defined in a following way:

```
- App 
    -> SharedFramework
    -> xcframeworkB
- SharedFramework, dynamic
    -> frameworkA: dynamic, mergeable
    -> xcframeworkB: dynamic, mergeable
    -> xcframeworkC: dynamic, mergeable
```

but such approach is **invalid** - the app should only link the `SharedFramework`, otherwise, `xcframeworkB` should be removed from the `SharedFramework`. 

There has to be a linter for such a thing, but I didn't have time yet to figure out how linters work in Tuist. So if you have any ideas how to easily write one, I might be able to open another MR.